narwhals/series.py

```diff
@@ -86,9 +91,9 @@ def shape(self) -> tuple[int]:
 
             We define a library agnostic function:
 
-            >>> def func(s_any):
-            ...     s = nw.from_native(s_any, series_only=True)
-            ...     return s.shape
+            >>> @nw.narwhalify
+            ... def func(s_any):
+            ...     return s_any.shape
 
```

Comment:
are we sure about `s_any` here?

- from the function's body perspective, `s: nw.Series`
- but, from the point of view of anyone reading the function, `s_any` is not a Narwhals series, it's their native object

Maybe we need a `IntoNarwhalsSeries` / `IntoNarwhalsDataFrame` / `IntoNarwhals/LazyFrame` / `IntoNarwhals` type aliases in `nw.typing`, so that people can annotate functions in a way that's legible to the user (i.e. "this function accepts anything which is convertible to Narwhals), but in the function body, their IDE can help them with type completion and all

Anyway, one for a separate PR, out-of-scope for this one, just thinking about it

Reply:
> from the function's body perspective, s: nw.Series

yes

> but, from the point of view of anyone reading the function, s_any is not a Narwhals series, it's their native object

more yes

Type hinting will need some more love all around I think. Here is exactly as you mention, from the user point of view, any native object can be passed, while the function will work with the narwhals objects - but that's taken care of by the decorator